proto/cacheclient.proto

```diff
@@ -108,3 +111,58 @@ message _DictionaryDeleteRequest {
 }
 
 message _DictionaryDeleteResponse {}
+
+message _SetFetchRequest {
+  bytes set_name = 1;
+}
+
+message _SetFetchResponse {
+  message _Found {
+    repeated bytes elements = 1;
+  }
+
+  message _Missing {}
+
+  oneof set {
+    _Found found = 1;
+    _Missing missing = 2;
+  }
+}
+
+message _SetUnionRequest {
+  bytes set_name = 1;
+  repeated bytes elements = 2;
+  uint64 ttl_milliseconds = 3;
+  bool refresh_ttl = 4;
+}
+
+message _SetUnionResponse {}
+
+message _SetDifferenceRequest {
+  // cache = request - stored
+  message _Minuend {
+    repeated bytes elements = 1;
+  }
+  // cache = stored - request
+  message _Subtrahend {
+    // Subtract a set of elements
+    message _Set {
+      repeated bytes elements = 1;
+    }
+    // Subtract the set's identity (itself) from itself - which deletes it.
+    message _Identity {}
+
+    oneof subtrahend_set {
+      _Set set = 1;
+      _Identity identity = 2;
+    }
+  }
+
+  bytes set_name = 1;
+  oneof difference {
+    _Minuend minuend = 2;
+    _Subtrahend subtrahend = 3;
+  }
+}
+
+message _SetDifferenceResponse {}
```

Comment:
@kvcache @malandis I know we decided that for read apis we would return the information if the set was found or missing and return the data if the set was found but for write apis which depend on the set to be present eg: SetDifference what should we return in the response? I think we should throw an exception here. And for the case where we want to delete the entire set (the set difference identity case) we will have to make it a no-op to be consistent with the dictionary delete all case. wdut?

Reply:
interesting. Looking forward to Mike's take.